packages/compose/src/hooks/use-focus-outside/index.ts

```diff
@@ -89,7 +89,7 @@ export default function useFocusOutside(
 
 	// Cancel blur checks on unmount.
 	useEffect( () => {
-		return () => cancelBlurCheck();
+		//return () => cancelBlurCheck();
```

Comment:
@WordPress/gutenberg-components In this PR unfocusing the `Popover` by clicking elsewhere _in the canvas_ causes the `onClose` callback supplied to `Popover` not to be called. This is because the timeout is cancelled by this effect when the component is unmounted.

However, if you click outside the canvas, such as on the List View, you will see that the timeout _is_ called.

I do not currently understand and would appreciate any assistance you can offer to debug this.

All I want to do is reliably trigger the `onClose` callback when the `Popover` is closed.

Reply:
Here's a screencapture showing what I mean:

https://github.com/WordPress/gutenberg/assets/444434/d0287838-ec46-4768-9eef-c53ee1cae45e

Reply:
> unfocusing the Popover by clicking elsewhere in the canvas causes the onClose callback supplied to Popover not to be called.

Mhh.. could that be related to the fact that, when clicking the canvas, the click is happening inside an `iframe` ?

Also cc'ing @stokesman who recently worked on related changes to the `Modal` component

Reply:
It's very odd. I thought it was something to do with that to because the set time out callback checks for `document.focus()` but that is not even being triggered because the timeout cancels _before_ it can be triggered.

If I had to guess I'd say the effect runs the timeout cancelling code before the next tick of the event loop (i.e. when the timeout's callback would be executed).

Reply:
I tried restoring the `return () => cancelBlurCheck();` line, and in my tests the popover is being closed when clicking on the canvas — I feel like I may be missing something on how to reproduce?

Re. the timeout callback not firing, another aspect that we should keep in mind is that, depending on the React tree, if the parent node of `Popover` gets unmounted, likely the timeout wouldn't have a chance to run — not sure if that's the case when canceling the addition of a new nav link.

Reply:
I tested after reverting 02bf15162b0291b8498b60a2cf0cb57b6ac98332 and got the same result as Marco. The popover closes from clicks inside the canvas. 🤷‍♂️

Reply:
Thanks for testing. Ok that's strange. I'll re-test. Perhaps there is a specific condition required to replicate.

Reply:
@jeryj Are you able to test this. If you cannot replicate then I guess it was some anomaly.

Reply:
I tested and I can't replicate it.

Reply:
Let's just assume it was something weird with my local setup. But at least if it does manifest we have a path for investigation.

Appreciate everyone jumping in here 🙇

Reply:
Actually sorry I tested without taking out the comment. Now I have removed it and this code is running again I can confirm this is still a problem. Sorry!

Reply:
It's still happening for me. There must be an extreme edge case here.

Reply:
If you're able to share some instructions on how to reproduce reliably, @WordPress/gutenberg-components may be able to take a look ?

Reply:
I'll try and do that this morning.

Reply:
@ciampo Here is [a walk through](https://d.pr/v/hdhso7) of how to reproduce this. I think it must have something to do with the iframe.

Reply:
Looking at your video, it definitely looks like it could be related to the `iframe`, confirming my [initial thoughts](https://github.com/WordPress/gutenberg/pull/57756#discussion_r1450780683).

Basically, my suspicion is that the `iframe` may acts as a "black hole" and doesn't forward the related browser event, which causes the `onClose` not to fire. Maybe you could do a quick test and render an invisible element in between the popover and the iframe which would act as way to intercept those mouse events? (I experienced something similar when working on https://github.com/WordPress/gutenberg/pull/55149).

Reply:
https://github.com/WordPress/gutenberg/assets/444434/8c44264f-3908-42f6-85a3-84fa7944265f

[I mirrored your implementation](https://github.com/WordPress/gutenberg/pull/57756/commits/8a39e54fc9a061a75276cb6614f8c582b018e0c4) as suggested and it does indeed work 🥳 

Should we consider making this part of the underlying `Popover` (as part of a followup)?

Reply:
It's not as easy — basically what we've been so far is the first step towards implementing "modality" (see this [great article](https://hidde.blog/dialog-modal-popover-differences/) for reference).

Doing it right is tricky, especially with a component like `Popover`. Ideally, at some point, we'd like to write a new version of `Popover` using `ariakit` which [would natively support modality](https://ariakit.org/components/popover). In the past few months we've been experimenting with rewriting a few components and I've been recently working with @mirka to standardise such process.

Finally, a word of caution regarding modality in general, but also the current "hack": all pointer events (clicks, hovers, etc) will be disabled while that popover is open. For example, to select another navigation link, the user will need two clicks instead of one — so this solution is not ideal.

The "right" solution, IMO, would be to understand how to allow interactions with `iframes` to work as expected and trigger the close button.  I know that the editor's `iFrame` already tries to do some [custom event forwarding / bubbling](https://github.com/WordPress/gutenberg/blob/33b6e641d6f3737a8880d22bad77cb4ee53be602/packages/block-editor/src/components/iframe/index.js#L75-L101), but I'm not familiar with it (cc @andrewserong who, if I'm not wrong, was also dealing with something related when working on drag and drop)

Reply:
I wonder if @ellatrix has some insight as well given her work on the iframed canvas?

Reply:
@getdave, I'm worried adding a div to close the popover is going to make this issue harder to unravel in the future. In the video you mention that if you click a link within the navigation, it does fire the callback. Does it also fire the callback if you click on another block or paragraph content?

Reply:
@getdave, I'm worried adding a div to close the popover is going to make this issue harder to unravel in the future. In the video you mention that if you click a link within the navigation, it does fire the callback. Does it also fire the callback if you click on another block or paragraph content?

Reply:
Created a followup here: https://github.com/WordPress/gutenberg/issues/58825